src/overlay.rs

```diff
@@ -286,21 +292,100 @@ impl OverlayMan {
     }
 }
 
-pub enum OverlayLayer<'a, G: xr::Graphics> {
+pub struct OverlayLayer<'a, G: xr::Graphics> {
+    /// Only ever None during next_chain_insert
+    layer: Option<OverlayLayerInner<'a, G>>,
+    color_bias_khr: Option<Box<xr::sys::CompositionLayerColorScaleBiasKHR>>,
+}
+
+impl<'a, G: xr::Graphics> OverlayLayer<'a, G> {
+    pub fn set_alpha(&mut self, alpha: f32) {
+        // only one instance is stored, so this would cause segfault due to UAF
+        debug_assert!(
+            self.color_bias_khr.is_none(),
+            "attempted to set_alpha on the same CompositorLayer twice!"
+        );
+
+        self.color_bias_khr = {
+            let mut payload = Box::new(xr::sys::CompositionLayerColorScaleBiasKHR {
+                ty: xr::StructureType::COMPOSITION_LAYER_COLOR_SCALE_BIAS_KHR,
+                next: std::ptr::null(),
+                color_bias: Default::default(),
+                color_scale: xr::Color4f {
+                    a: alpha,
+                    ..Default::default()
+                },
+            });
+
+            let payload_ptr = payload.as_mut() as *mut _ as *mut xr::sys::BaseInStructure;
+            unsafe { self.next_chain_insert(payload_ptr) };
+
+            Some(payload)
+        };
+    }
+
+    /// Insert the given item as the first element in the next chain.
+    /// `item` must be a non-null pointer to a valid XrBaseInStructure object
+    ///
+    /// SAFETY: For lifetime guarantees, store item in Box inside CompositorLayer.
+    unsafe fn next_chain_insert(&mut self, item: *mut xr::sys::BaseInStructure) {
+        let new_elem = item.as_mut().unwrap();
+        self.layer = Some(match self.layer.take().unwrap() {
+            OverlayLayerInner::Quad(quad) => {
+                let mut raw = quad.into_raw();
+                new_elem.next = raw.next as _;
+                raw.next = item.as_ref().unwrap() as *const _ as *const _;
+                OverlayLayerInner::Quad(xr::CompositionLayerQuad::<'a, G>::from_raw(raw))
```

Comment:
Could the turbofish be elided?
```suggestion
                OverlayLayerInner::Quad(xr::CompositionLayerQuad::from_raw(raw))
```